classes/userGroup/Repository.php

```diff
@@ -0,0 +1,469 @@
+     * @param array $props A key/value array with the new data to validate
+     * @param array $allowedLocales The context's supported submission locales
+     * @param string $primaryLocale The submission's primary locale
+     *
+     * @return array A key/value array with validation errors. Empty if no errors
+     */
+    public function validate($userGroup, $props, $allowedLocales, $primaryLocale)
+    {
+        $schemaService = Services::get('schema');
+
+        $validator = ValidatorFactory::make(
+            $props,
+            $schemaService->getValidationRules(PKPSchemaService::SCHEMA_USER_GROUP, $allowedLocales)
+        );
+
+        // Check required fields
+        ValidatorFactory::required(
+            $validator,
+            $userGroup,
+            $schemaService->getRequiredProps(PKPSchemaService::SCHEMA_USER_GROUP),
+            $schemaService->getMultilingualProps(PKPSchemaService::SCHEMA_USER_GROUP),
+            $allowedLocales,
+            $primaryLocale
+        );
+
+        // Check for input from disallowed locales
+        ValidatorFactory::allowedLocales($validator, $schemaService->getMultilingualProps(PKPSchemaService::SCHEMA_USER_GROUP), $allowedLocales);
+
+        $errors = [];
+        if ($validator->fails()) {
+            $errors = $schemaService->formatValidationErrors($validator->errors());
+        }
+
+        HookRegistry::call('UserGroup::validate', [$errors, $userGroup, $props, $allowedLocales, $primaryLocale]);
+
+        return $errors;
+    }
+
+    /**
+     * @copydoc \PKP\services\entityProperties\EntityWriteInterface::add()
+     */
+    public function add(UserGroup $userGroup): int
+    {
+        $userGroupId = $this->dao->insert($userGroup);
+        $userGroup = Repo::userGroup()->get($userGroupId);
+
+        HookRegistry::call('UserGroup::add', [$userGroup]);
+
+        return $userGroup->getId();
+    }
+
+    /**
+     * @copydoc \PKP\services\entityProperties\EntityWriteInterface::edit()
+     */
+    public function edit(UserGroup $userGroup, array $params)
+    {
+        $newUserGroup = Repo::userGroup()->newDataObject(array_merge($userGroup->_data, $params));
+
+        HookRegistry::call('UserGroup::edit', [$newUserGroup, $userGroup, $params]);
+
+        $this->dao->update($newUserGroup);
+
+        Repo::userGroup()->get($newUserGroup->getId());
+    }
+
+    /**
+     * @copydoc \PKP\services\entityProperties\EntityWriteInterface::delete()
+     */
+    public function delete(UserGroup $userGroup)
+    {
+        HookRegistry::call('UserGroup::delete::before', [$userGroup]);
+
+        $this->dao->delete($userGroup);
+
+        HookRegistry::call('UserGroup::delete', [$userGroup]);
+    }
+
+    /**
+    * Delete all user groups assigned to a certain context by contextId
+    *
+    * @param int $contextId
+    */
+    public function deleteByContextId($contextId)
+    {
+        // I may have to use transactions here
```

Comment:
I'd like to figure out a consistent policy on transactions before we start to use them. We can't nest transactions, so applying them at one level means we can't apply them on another level.

Reply:
Transactions can be nested using a [savepoint](https://www.tutorialspoint.com/mysql/mysql_savepoint.htm).
I was checking one issue in Laravel, and looks like they've got it implemented (https://github.com/laravel/framework/issues/1823#issuecomment-148954287), but better to test 😂